[INF-6633] Add retry logic to client

martin-ably opened the pull request:
Adds retry logic by implementing hashicorp/go-retryablehttp instead of
net/http.
Additionally fixes a handful of failing tests due to updates performed by website to controlAPI.

I expect this change should be mostly backwards-compatible. The only concern is behavioural, in cases where downstream applications have implemented their own retry logic it may clash with that, though the client should preempt that by itself since it shouldn't error out until it's finished retrying, so it may be an entirely transparent change overall.

surminus reviewed (approved):
Looks good to me 👍

lmars commented on client.go:
```diff
@@ -40,24 +43,94 @@ type Client struct {
+// WithBackoff sets a custom backoff strategy function.
+// The function receives min/max wait times, the attempt number (starting at 0),
+// and the response, and returns the duration to wait before the next retry.
+func WithBackoff(backoff retryablehttp.Backoff) ClientOption {
```
I'd probably type alias `retryablehttp.Backoff` so that the `github.com/hashicorp/go-retryablehttp` package isn't in the public API.

martin-ably replied:
Ok, I've added `type Backoff = retryablehttp.Backoff` and `type CheckRetry = retryablehttp.CheckRetry` now which should fix this.

lmars commented on client_test.go:
```diff
@@ -20,26 +23,31 @@ func TestMain(m *testing.M) {
+// skipIfNoToken skips a test if ABLY_ACCOUNT_TOKEN is not set.
+func skipIfNoToken(t *testing.T) {
+	if token == "" {
+		t.Skip("ABLY_ACCOUNT_TOKEN not set, skipping integration test")
```
Does this have the potential to silently skip all integration tests if there's a configuration error and `ABLY_ACCOUNT_TOKEN` becomes unset in CI?

martin-ably replied:
It would skip them.
I personally consider it exceedingly unlikely that the token would become unset (and that would arguably be indicative of more significant problems in our config change process), but I can switch out the account token check to check for a different flag like "SKIP_INTEGRATION" if that would help?
It was added mainly as a way to make testing, when I didn't need to check functions that call the API, quicker

lmars replied:
Yeh I think I'd prefer a flag which explicitly skips them that can be used locally.

lmars commented on namespaces_test.go:
```diff
@@ -79,6 +79,12 @@ func TestNamespaces(t *testing.T) {
+	// comparison but after the request is sent as it can't be explicitly
+	// enabled when conflation is enabled in the request.
+	namespace.BatchingEnabled = true
+	namespace.BatchingInterval = Interval(1000)
```
What is the purpose of this test? It seems odd that this is asserting what I assume is a default value of `batchingInterval=1000` which could change within the system and break this test.

martin-ably replied:
Honestly, because the tests are entirely uncommented, I'm presuming that the whole test those extra lines are in is for testing whether conflation is set correctly when set to true in the input struct.
```
namespace = Namespace{
	ID:                 namespace.ID,
	Authenticated:      true,
	Persisted:          true,
	PersistLast:        true,
	PushEnabled:        true,
	TlsOnly:            true,
	ExposeTimeserial:   true,
	ConflationEnabled:  true,
	ConflationInterval: Interval(1000),
	ConflationKey:      "test",
}

n, err = client.UpdateNamespace(app.ID, &namespace)
assert.NoError(t, err)
// API seems to auto-enable batching when conflation is enabled
// so we need to set it to true in our input before we do the
// comparison but after the request is sent as it can't be explicitly
// enabled when conflation is enabled in the request.
namespace.BatchingEnabled = true
namespace.BatchingInterval = Interval(1000)
assert.Equal(t, namespace, n)
```
However, the API response when conflation is set also returns Batching as being set with an interval that the system itself seems to decide to set, and no clear documentation that I could find on why, meaning I added those lines to reflect the state of the API so that the test is still correct, rather than just removing the test or bypassing it.
If you remove the two lines I added you get the following error:
```
2025/12/23 10:32:33 [DEBUG] PATCH https://control.ably.net/v1/apps/HinQhw/namespaces/test-namespace-16341787099872010163
    namespaces_test.go:88:
                Error Trace:    namespaces_test.go:88
                Error:          Not equal:
                                expected: control.Namespace{ID:"test-namespace-16341787099872010163", Authenticated:true, Persisted:true, PersistLast:true, PushEnabled:true, TlsOnly:true, ExposeTimeserial:true, BatchingEnabled:false, BatchingInterval:(*int)(nil), ConflationEnabled:true, ConflationInterval:(*int)(0xc000013020), ConflationKey:"test"}
                                actual  : control.Namespace{ID:"test-namespace-16341787099872010163", Authenticated:true, Persisted:true, PersistLast:true, PushEnabled:true, TlsOnly:true, ExposeTimeserial:true, BatchingEnabled:true, BatchingInterval:(*int)(0xc0003326b0), ConflationEnabled:true, ConflationInterval:(*int)(0xc0003326b8), ConflationKey:"test"}

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -8,4 +8,4 @@
                                  ExposeTimeserial: (bool) true,
                                - BatchingEnabled: (bool) false,
                                - BatchingInterval: (*int)(<nil>),
                                + BatchingEnabled: (bool) true,
                                + BatchingInterval: (*int)(1000),
                                  ConflationEnabled: (bool) true,
                Test:           TestNamespaces
```
Which indicates that for some reason the API itself is changing the values, and thus the test falls through since we aren't setting batching in the input but using the input to validate the returned data. Meanwhile, if you set batching enabled and the batching interval in the input instead, the test also fails with an error about not being able to set both batching and conflation simultaneously in the input.
```
2025/12/23 10:35:37 [DEBUG] PATCH https://control.ably.net/v1/apps/qTapCw/namespaces/test-namespace-7539925423458520038
    namespaces_test.go:83:
                Error Trace:    namespaces_test.go:83
                Error:          Received unexpected error:
                                /apps/qTapCw/namespaces/test-namespace-7539925423458520038: Invalid resource: code 40022: status code: 422 see: https://help.ably.io/error/40022
                                  base:
                                    Server-side batching and message conflation are mutually exclusive. You cannot configure both using the same namespace, or channel.
                Test:           TestNamespaces
    namespaces_test.go:90:
                Error Trace:    namespaces_test.go:90
                Error:          Not equal:
                                expected: control.Namespace{ID:"test-namespace-7539925423458520038", Authenticated:true, Persisted:true, PersistLast:true, PushEnabled:true, TlsOnly:true, ExposeTimeserial:true, BatchingEnabled:true, BatchingInterval:(*int)(0xc00020eff8), ConflationEnabled:true, ConflationInterval:(*int)(0xc00020eff0), ConflationKey:"test"}
                                actual  : control.Namespace{ID:"", Authenticated:false, Persisted:false, PersistLast:false, PushEnabled:false, TlsOnly:false, ExposeTimeserial:false, BatchingEnabled:false, BatchingInterval:(*int)(nil), ConflationEnabled:false, ConflationInterval:(*int)(nil), ConflationKey:""}

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,14 +1,14 @@
                                 (control.Namespace) {
                                - ID: (string) (len=34) "test-namespace-7539925423458520038",
                                - Authenticated: (bool) true,
                                - Persisted: (bool) true,
                                - PersistLast: (bool) true,
                                - PushEnabled: (bool) true,
                                - TlsOnly: (bool) true,
                                - ExposeTimeserial: (bool) true,
                                - BatchingEnabled: (bool) true,
                                - BatchingInterval: (*int)(1000),
                                - ConflationEnabled: (bool) true,
                                - ConflationInterval: (*int)(1000),
                                - ConflationKey: (string) (len=4) "test"
                                + ID: (string) "",
                                + Authenticated: (bool) false,
                                + Persisted: (bool) false,
                                + PersistLast: (bool) false,
                                + PushEnabled: (bool) false,
                                + TlsOnly: (bool) false,
                                + ExposeTimeserial: (bool) false,
                                + BatchingEnabled: (bool) false,
                                + BatchingInterval: (*int)(<nil>),
                                + ConflationEnabled: (bool) false,
                                + ConflationInterval: (*int)(<nil>),
                                + ConflationKey: (string) ""
                                 }
                Test:           TestNamespaces
```

Ergo, I'm not sure where I'd get the values for batching from, when enabling conflation, that are being set by the API, because they don't come from the input, and if for example the batching interval is derived from the conflation interval, then the test is just as much at risk of breaking in the future because the equation for deriving the batching interval could also arbitrarily change...

lmars replied:
> I'm presuming that the whole test those extra lines are in is for testing whether conflation is set correctly when set to true in the input struct.

Ok, if that is the intention of the test, then perhaps it should be asserting just the value of `ConflationEnabled` on the output struct, rather than asserting the whole structure with `assert.Equal`?

martin-ably committed:
Add retry logic to httpClient

Adds retry logic by implementing hashicorp/go-retryablehttp instead of
net/http.

martin-ably committed:
Fix failing tests

Fixes the failing HttpAzureFunction rule test. Since we don't seem
to have a proper endpoint for testing this and the API is actually
checking now, I'm using one of the app IDs found in the website tests.

Also, switch from test.com to example.com as the latter actually
resolves dns records.

Fixes namespace test by modifying the struct we are comparing the
response against as the batching value seems to be enabled by default in
the response when conflation is enabled, but we cannot have both enabled
at the same time explicitly in the input.
